checkptys.8
.TH checkptys 8
.SH NAME
checkptys \- check pseudo-terminal use and pty installation
.SH SYNOPSIS
.B checkptys
.SH DESCRIPTION
.B checkptys
produces a report on the entire pseudo-terminal system,
including users currently logged on,
the protection of special files such as
/etc/utmp,
and the setup of the
.B pty
package.

It is assumed that
.B checkptys
will be run with the same effective uid as
.B pty.
You may want to run it from the daily cron script.
.SH BUGS
This man page is woefully incomplete.
.SH "SEE ALSO"
pty(1)


CHANGES
2/9/92: pty 4.0.
2/9/92: Fixed up documentation, hopefully for the last time.
1/29/92: Changed INSTALL.c to report chown root for INSTALL root. Tnx IS.
1/27/92: Added ABRT->IOT ifndef in sigdfl.c.
1/27/92: Changed sigdfl.c to zero the sigvecs rather than setting sv_flags.
1/16/92: pty 4.0, second gamma.
1/13/92: Does it work on an Iris? Nobody knows. 
1/13/92: Added IRIS_UGH_PTYS to ptyget.c. Tnx IRIS, I think.
1/13/92: Changed fatal to warning for tctpgrp failure in ptyslave.c. Tnx IRIS.
1/13/92: ifndef'ed SYSV for ut_host in ptylogs.c. Tnx IRIS.
1/13/92: Added config/sysv.h to distribution. Tnx IRIS.
1/13/92: Added PTYLIBS, noted its use in README. Tnx IRIS.
1/13/92: Replaced CCOPTMZ with OPTMZ.
1/13/92: Noted in CALLS that TTY_TERMIO seems to work. Tnx IRIS.
1/13/92: Adapted tty_initmodes for TTY_TERMIO. Ugh, ugh. Tnx IRIS.
1/13/92: Adapted tty_mungemodes for TTY_TERMIO. Ugh. Tnx IRIS.
1/13/92: Adapted tty_modifymodes for TTY_TERMIO (poorly). Tnx IRIS.
1/13/92: Adapted tty_getmodes and tty_setmodes for TTY_TERMIO. Tnx IRIS.
1/13/92: In ptytty.c used ifdefs for FIONREAD and TIOCEXCL. Tnx IRIS.
1/13/92: Adapted ptytty.h for TTY_TERMIO. Tnx IRIS.
1/13/92: Added TTY_TERMIO option to config/ttyopts.h. Tnx IRIS.
1/13/92: Gave in and made tty_copymodes do a struct copy.
1/13/92: Fixed typo in SECURITY. Tnx KS.
1/13/92: Fixed typo in pty-basic.1. Tnx KS.
1/12/92: Replaced <sgtty.h> with <sys/ioctl.h> in ptytty.h. Tnx IRIS.
1/12/92: Added chmod 700 for all shell scripts in Makefile.
1/12/92: Removed <sys/timeb.h> from wall.c and write.c. Tnx IRIS.
1/12/92: In separate effort got it working on an IRIS. See above.
12/22/91: pty 4.0, first gamma! (Still have to integrate BSD 4.4 version.)
12/22/91: Changed various things in the documentation.
12/22/91: Fixed exit code reporting for, e.g., DIE_NOPTYS.
12/22/91: Changed POSIX to POSIX_SILLINESS. See config/posix.h.
12/20/91: Added CCOPTMZ in Makefile for radixsort.
12/19/91: Added BLURB.
12/19/91: Added TESTS.
12/19/91: Cleaned up README.
12/19/91: Added WTF, noted SunOS 4.0.3 forcefg bug in README.
12/18/91: Fixed bug in sessmenu's sesslist parsing.
12/17/91: Fixed up documentation a bit; referred to checkptys in README.
12/16/91: Added CPP define to Makefile.
12/15/91: Added -xi.
12/15/91: Started machine list in Makefile.
12/15/91: Added EWOULDBLOCK in ptysecure.c for Domain/OS.
12/14/91: Added target-list target.
12/14/91: Added lint target.
12/13/91: Changed master so that it continues child on any reconnect. Tnx SS.
11/24/91: Changed master to run only as effective uid.
11/18/91: Added TERM, USR1, USR2 ignoring to master. Tnx SS.
11/9/91: Added \fP in syntax sections in pty-opts.1, pty.1. Tnx SS.
11/9/91: Added <sys/types.h> before <utmp.h> in {users,who}.c. Domain/OS idiocy.
11/9/91: Fixed dependencies for CHECKCONF and INSTALL.
11/3/91: Added proper .h dependencies in Makefile (finally).
10/27/91: Added exit(0) at end of sesslist.c. Tnx SS.
10/27/91: Fixed ptysecure.c---it didn't undo chmod for some setups. Tnx SS.
10/27/91: Added make clean. Tnx SS.
10/27/91: Added checkptys.8.
10/27/91: Changed checkptys to be installed non-setuid.
10/26/91: Added pty.1, pty-basic.1, pty-opts.1, sess.1, condom.1.
10/26/91: Tossed old man pages.
10/26/91: Added DIE_EXIST for ptysigler.c.
10/26/91: Added DIE_ELSE for ptysigler.c.
10/26/91: Fixed ptymain to die(SETMODES) in an impossible situation.
10/26/91: Changed a few die() values in ptymain.c.
10/25/91: Need to add *.h to Makefile...
10/25/91: Removed last traces of old getopt schizophrenia.
10/25/91: Changed *rotate to preserve modes of log files.
10/25/91: Added sessnowinit, scnowinit.
10/25/91: Added sessrotate, sclogrotate.
10/25/91: Fixed off-by-one on remotelen.
10/25/91: Changed wtmprotate to depend on config/wtmpfile.h properly.
[log is not complete before this point]
10/6/91: Added -xsS.
9/30/91: pty 4.0, second beta.


checkptys.c
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/file.h>
#include "config/devmty.h"
#include "config/devsty.h"
#include "config/posix.h" /* XXX: why? */
#include "config/ptybin.h"
#include "config/ptydir.h"
#include "config/ptyext.h"
#include "config/ptygroup.h"
#include "config/ptymodes.h"
#include "config/ptyopts.h"
#include "config/sessconnfile.h"
#include "config/sessfile.h"
#include <utmp.h>
#include "config/utmpfile.h"
#include "config/wtmpfile.h"

char ptybin[] = PTYBIN;
char ptydir[] = PTYDIR;
char sessconnnow[] = SESSCONNNOW_FILE;
char sessconnlog[] = SESSCONNLOG_FILE;
char sessnow[] = SESSNOW_FILE;
char sesslog[] = SESSLOG_FILE;
char utmp[] = UTMP_FILE;
char wtmp[] = WTMP_FILE;

char devmty[sizeof(DEVMTY) + 5] = DEVMTY;
char devsty[sizeof(DEVSTY) + 5] = DEVSTY;
char pty1[] = PTYEXT1;
char pty2[] = PTYEXT2;

void aack(s,t)
char *s;
char *t;
{
 printf("aack! %s: %s\n",s,t);
}

void checkbin(s,level)
char *s;
int level; /* 0 script 1 normal 2 setgid 3 setuid */
{
 struct stat st;

 if (stat(s,&st) == -1)
  {
   aack("executable does not exist",s);
   return;
  }
 if (!(st.st_mode & 001))
   aack("program isn't world-executable",s);
 switch(st.st_mode & 06000)
  {
   case 0:
     if (level == 2)
       aack("program should be setgid but isn't",s);
     if (level == 3)
       aack("program should be setuid but isn't",s);
     break;
   case 06000:
   case 04000:
     if (level == 0)
       aack("SHELL SCRIPT IS SETUID! FIX IMMEDIATELY!",s);
     if (level == 1)
       aack("program is SETUID, shouldn't be",s);
     if (level == 2)
       aack("program is SETUID, should be setgid instead",s);
     break;
   case 02000:
     if (level == 0)
       aack("SHELL SCRIPT IS SETGID! FIX IMMEDIATELY!",s);
     if (level == 1)
       aack("program is setgid, shouldn't be",s);
     if (level == 3)
       aack("program is setgid, should be setuid instead",s);
     break;
   default:
     aack("computer doesn't understand binary arithmetic",s);
  }
 if (st.st_mode & 002)
   aack("program is WORLD-writable",s);
 if ((level == 2) && (st.st_gid != PTYGROUP))
   aack("program should be setgid to tty group but isn't",s);
}

void testpath(s)
char *s;
{
 char *t;
 struct stat st;
 char old;

 if (*s != '/')
  {
   aack("path does not start with a slash",s);
   return;
  }
 t = s + 1;
 for (;;)
  {
   if ((*t == '/') || (*t == 0))
    {
     old = *t;
     *t = 0;
     if (stat(*s ? s : "/",&st) == -1)
       aack("cannot stat component of path",*s ? s : "/");
     else
      {
       if (st.st_mode & 022)
	 aack("component directory is WORLD- and group-writable",*s ? s : "/");
       else if (st.st_mode & 002)
	 aack("component directory is WORLD-writable",*s ? s : "/");
       else if (st.st_mode & 020)
	 aack("component directory is group-writable",*s ? s : "/");
      }
     *t = old;
    }
   if (!*t)
     break;
   ++t;
  }
}

void testpty(fnm,fns,unusedptyowner)
char *fnm;
char *fns;
int unusedptyowner;
{
 struct stat stm;
 struct stat sts;
 int fdm;

 if (stat(fnm,&stm) == -1)
   return;
 if (stat(fns,&sts) == -1)
  {
   aack("pty master has no corresponding slave",fnm);
   return;
  }

 fdm = open(fnm,O_RDWR);
 if (stm.st_mode & 0777 != 0666)
  {
   aack("pty master has weird mode",fnm);
  }
 if (fdm == -1)
  {
   printf("pty master %s in use, slave owned by uid %d\n",fnm,sts.st_uid);
   if (sts.st_mode & 006)
     aack("pty slave allows WORLD access",fns);
   if (sts.st_gid != PTYGROUP)
     aack("pty slave group does not match standard tty group",fns);
   if (sts.st_mode & 020)
     printf("pty slave %s messages on\n",fns);
   if (sts.st_mode & 0100)
     printf("pty slave %s biff on\n",fns);
  }
 else
  {
   if (sts.st_uid != unusedptyowner)
     aack("unused pty slave not owned by standard unused tty owner",fns);
   close(fdm);
  }
}

main()
{
 int p1;
 int p2;

 printf("Testing main pty directory %s...\n",ptydir);
 testpath(ptydir);
 printf("Testing utmp file %s...\n",utmp);
 testpath(utmp);
 printf("Testing wtmp file %s...\n",wtmp);
 testpath(wtmp);
 printf("Testing session log file %s...\n",sesslog);
 testpath(sesslog);
 printf("Testing current session file %s...\n",sessnow);
 testpath(sessnow);
 printf("Testing session-connection log file %s...\n",sessconnlog);
 testpath(sessconnlog);
 printf("Testing current session-connection file %s...\n",sessconnnow);
 testpath(sessconnnow);
 printf("Testing pty binary directory %s...\n",ptybin);
 testpath(ptybin);

 for (p1 = 0;pty1[p1];++p1)
   for (p2 = 0;pty2[p2];++p2)
    {
     devmty[sizeof(DEVMTY) - 1] = pty1[p1];
     devmty[sizeof(DEVMTY)] = pty2[p2];
     devsty[sizeof(DEVSTY) - 1] = pty1[p1];
     devsty[sizeof(DEVSTY)] = pty2[p2];
     testpty(devmty,devsty,geteuid());
    }

 printf("Checking for actual pty-related binaries in %s...\n",ptybin);
 if (chdir(ptybin) == -1)
  {
   aack("cannot switch to pty binary directory",ptybin);
  }
 else
  {
   checkbin("argv0",1);
   checkbin("biff",1);
   checkbin("checkptys",1);
   checkbin("condom",0);
   checkbin("ctrlv",1);
   checkbin("disconnect",3);
   checkbin("excloff",1);
   checkbin("exclon",1);
   checkbin("lock",1);
   checkbin("mesg",1);
   checkbin("nobuf",0);
   checkbin("pty",3);
   checkbin("reconnect",3);
   checkbin("script",0);
   checkbin("script.tidy",0);
   checkbin("sess",0);
   checkbin("sesskill",3);
   checkbin("sesslist",3);
   checkbin("sessmenu",1);
   checkbin("sessname",3);
   checkbin("sesswhere",1);
   checkbin("sesswho",1);
   checkbin("tiocsti",1);
   checkbin("tplay",1);
   checkbin("trecord",1);
   checkbin("tscript",0);
   checkbin("tty",1);
   checkbin("ttydetach",1);
   checkbin("ttyprotect",0);
   checkbin("users",1);
   checkbin("utmpinit",1);
   checkbin("waitfor",1);
   checkbin("wall",2);
   checkbin("who",1);
   checkbin("whoami",1);
   checkbin("write",2);
   checkbin("wtmprotate",0);
   checkbin("sessrotate",0);
   checkbin("sclogrotate",0);
   checkbin("sessnowinit",0);
   checkbin("scnowinit",0);
  }

 exit(0);
}
